README.md
# vending_machine


history.txt
Coca	may1	2016-5-18 11:43:7
Coca	may1	2016-5-18 11:45:10
Coca	may1	2016-5-18 11:45:10
Coca	may1	2016-5-18 11:45:10
Coca	may1	2016-5-18 11:46:37
Coca	may1	2016-5-18 11:47:44
Coca	may1	2016-5-18 11:49:36
Coca	may1	2016-5-18 11:50:43
may1	Coca	2016-5-18 11:52:9
may1	Coca	2016-5-18 18:31:40
may1	Coca	2016-5-18 18:31:40
may1	Coca	2016-5-18 18:34:48
may1	Coca	2016-5-18 18:35:57
may1	Coca	2016-5-18 18:35:57
may2	Coca	2016-5-18 18:35:57
may1	Coca	2016-5-18 18:38:8
may1	Coca	2016-5-18 18:39:37
may1	Coca	2016-5-18 18:43:53
may1	Coca	2016-5-18 18:44:58
may1	Coca	2016-5-18 18:44:58
may1	Coca	2016-5-18 18:47:41
may1	Pepsi	2016-5-18 18:47:41
may1	Coca	2016-5-18 18:49:53
may1	Coca	2016-5-18 18:53:6
may2	Coca	2016-5-18 18:56:8
may2	Coca	2016-5-18 18:57:31
may2	Coca	2016-5-18 18:59:15
may2	Coca	2016-5-18 18:59:15
may1	Coca	2016-5-18 18:59:15
may1	Coca	2016-5-18 19:9:46
may1	Coca	2016-5-18 19:12:13
may1	Coca	2016-5-18 19:13:42
may1	Coca	2016-5-18 19:18:40
may1	Coca	2016-5-18 19:21:23
may1	Coca	2016-5-18 19:23:20
may1	Coca	2016-5-18 19:25:2
may1	Coca	2016-5-18 19:26:21
may1	Coca	2016-5-18 19:32:29
may1	Coca	2016-5-18 22:18:8
may1	Coca	2016-5-18 22:20:17
may1	Coca	2016-5-18 22:26:10
may1	Coca	2016-5-18 22:31:21
may1	Coca	2016-5-18 22:33:44
may1	Coca	2016-5-18 22:40:57
may1	Coca	2016-5-18 22:48:3
may1	Coca	2016-5-18 22:56:5
may1	Coca	2016-5-18 23:0:53
may1	Coca	2016-5-18 23:14:50
may2	Coca	2016-5-18 23:14:50
may1	Coca	2016-5-18 23:17:15
may2	Pepsi	2016-5-18 23:17:15
may1	Coca	2016-5-18 23:22:57
may1	Coca	2016-5-18 23:24:0
may1	Coca	2016-5-18 23:26:5
may1	Coca	2016-5-18 23:38:34
may1	Coca	2016-5-18 23:38:34
may1	Coca	2016-5-18 23:38:34
may1	Coca	2016-5-18 23:39:45
may1	Coca	2016-5-18 23:39:45
may1	Coca	2016-5-18 23:41:27
may1	Coca	2016-5-18 23:41:27
may1	Coca	2016-5-18 23:41:27
may1	Coca	2016-5-18 23:41:27
may2	Coca	2016-5-18 23:41:27
may1	Coca	2016-5-18 23:46:18
may1	Coca	2016-5-18 23:46:18
may2	Coca	2016-5-18 23:46:18
may1	Coca	2016-5-18 23:47:35
may1	Coca	2016-5-18 23:47:35
may1	Coca	2016-5-18 23:47:35
may1	Pepsi	2016-5-18 23:47:35
may2	Coca	2016-5-18 23:47:35
may2	Coca	2016-5-18 23:47:35
may2	Coca	2016-5-18 23:47:35
may1	Coca	2016-5-18 23:47:35
may2	Pepsi	2016-5-18 23:47:35
may1	Coca	2016-5-18 23:49:35
may2	Coca	2016-5-18 23:49:35
may1	Coca	2016-5-18 23:51:19
may2	Coca	2016-5-18 23:51:19
may1	Coca	2016-5-18 23:54:8
may1	Coca	2016-5-18 23:54:8
may1	Coca	2016-5-18 23:54:8
may1	Coca	2016-5-18 23:54:8
may1	Coca	2016-5-18 23:54:8
may1	Coca	2016-5-18 23:56:49
may1	Coca	2016-5-18 23:56:49
may2	Coca	2016-5-18 23:56:49
may2	Coca	2016-5-18 23:56:49
may2	Coca	2016-5-18 23:56:49
may2	Coca	2016-5-18 23:56:49
may2	Coca	2016-5-18 23:56:49
may1	Coca	2016-5-18 23:56:49
may2	Coca	2016-5-18 23:56:49


client.c
/*
    C ECHO client example using sockets
*/
#include <stdio.h> //printf
#include <string.h>    //strlen
#include <sys/socket.h>    //socket
#include <arpa/inet.h> //inet_addr

char* getBranchName();
char* getMachineName();
 
int main(int argc , char *argv[]){
    int sock,equipMain;
    struct sockaddr_in server;
    char *branchName , *machineName; 
    char server_reply[2000];
    char message[64];
    char reply[64];
    char success[]="Successfully!";
    char machineError[]="Wrong Machine";
    //Create socket
    sock = socket(AF_INET , SOCK_STREAM , 0);
    if (sock == -1)
    {
        printf("Could not create socket");
    }
    puts("Socket created");
     
    server.sin_addr.s_addr = inet_addr("127.0.0.1");
    server.sin_family = AF_INET;
    server.sin_port = htons( 8888 );
 
    //Connect to remote server
    if ((equipMain=connect(sock , (struct sockaddr *)&server , sizeof(server))) < 0){
        perror("connect failed. Error");
        return 1;
    }
     
    puts("Connected\n");
     
    //keep communicating with server
    for(;;){
        branchName=getBranchName();strcpy(message,branchName);
        machineName=getMachineName(); strcat(message,",");
        strcat(message,machineName);strcat(message,"\n");
        
        //Send some data
        if(send(sock,message,strlen(message)+1,0)<0){
            perror("Can't send");
        }
        //Receive a reply from the server
        if( recv(sock , reply , 2000 , 0) < 0){
            puts("recv failed");
            break;
        }
        if(strcmp(reply,machineError)==0){
            puts(reply);
            break;
        }
        puts(reply);
    }
        close(sock);
     
      return 0;
}

char* getBranchName(){
    char* branch;
    branch=(char*)malloc(32*sizeof(char));
    printf("Enter Branch Name: ");
    scanf("%s" , branch);
    return branch;
}
char* getMachineName(){
    char* machineName;
    machineName=(char*)malloc(32*sizeof(char)); 
    printf("Enter Machine Name:");
    scanf("%s", machineName);
    return machineName;
}

update.c
 
#include <stdio.h>
#include <string.h>    //strlen
#include <sys/socket.h>
#include <arpa/inet.h> //inet_addr
#include <unistd.h>    //write
#include <time.h>
#include <stdlib.h>

void waitFor (unsigned int secs) {
    unsigned int retTime = time(0) + secs;   // Get finishing time.
    while (time(0) < retTime);               // Loop until it arrives.
}



int main(){
	char *branch, *machine,*amount;
	char character;
	char info[1024],invent[4];
	int number;
	FILE*inventFile;
	unsigned int secs=5;
	if((inventFile=fopen("inventory.txt","r+"))==NULL){
          perror("error:");
    }
	//for(;;){
		void waitFor (secs);
	    while(1){
	            fgets(info,2000,inventFile);
	            if(feof(inventFile)) break;
	            machine=strtok(info,",");printf("machine:%s\n",machine);
	            branch=strtok(NULL,",");
	            amount=strtok(NULL,"\n ");number=atoi(amount);
	            
	            if(number<=3){
	                sprintf(invent,"%d",number+10);
	                fseek(inventFile,-4,SEEK_CUR);
	                while((character=fgetc(inventFile))!='\n'){
	                        if(isdigit((int)character)){
	                            fseek(inventFile,-1,SEEK_CUR);
	                            fprintf(inventFile,"%s",invent);
	                            fseek(inventFile,1,SEEK_CUR);
	                            break;
	                        }
	                }
	            }
	        }
    //}
	 fclose(inventFile);
}

server.c
/*
    C socket server example
*/
 
#include <stdio.h>
#include <string.h>    //strlen
#include <sys/socket.h>
#include <arpa/inet.h> //inet_addr
#include <unistd.h>    //write
#include <time.h>
#include <stdlib.h>

typedef struct Inventory_t{
    char machineName[32];
    char branchName[32];
    int  amount;
}Inventory;


void initInventFile();
void waitFor (unsigned int secs);

int main(int argc , char *argv[]){
    int socket_desc , client_sock , c , read_size, confirm=0;
    int counter;
    int number;
    struct sockaddr_in server , client;
    char client_message[2000];
    char machineError[]="Wrong Machine"; char success[]="Successfully!";
    char *branchName, *machineName;
    FILE* inventFile, *historyFile;
    Inventory* inventory;
    char back=8,character;
    char invent[4];

    unsigned int secs=5;

    time_t t = time(NULL);
    struct tm tm;

    char info[1024];char*machine;char*branch;char*amount;
    // Create file pointer
    
    
    //Create socket
    socket_desc = socket(AF_INET , SOCK_STREAM , 0);
    if (socket_desc == -1)
    {
        printf("Could not create socket");
    }
    puts("Socket created");
     
    //Prepare the sockaddr_in structure
    server.sin_family = AF_INET;
    server.sin_addr.s_addr = INADDR_ANY;
    server.sin_port = htons( 8888 );
     
    //Bind
    if( bind(socket_desc,(struct sockaddr *)&server , sizeof(server)) < 0)
    {
        //print the error message
        perror("bind failed. Error");
        return 1;
    }
    puts("bind done");
     
    //Listen
    listen(socket_desc , 3);
     
    //Accept and incoming connection
    puts("Waiting for incoming connections...");
    c = sizeof(struct sockaddr_in);
    int pid;

    //test
    
    if((inventFile=fopen("inventory.txt","r+"))==NULL){
          perror("error:");
    }
   
        fclose(inventFile);
    //test
     
    //accept connection from an incoming client
    for(;;){
        client_sock = accept(socket_desc, (struct sockaddr *)&client, (socklen_t*)&c);
        pid=fork();
        if(pid<0){
            close(client_sock);
            continue;
        }else if(pid>0){
            close(client_sock);
            counter++;
            printf("here2\n");
            continue;
        }else{
                if (client_sock < 0){
                    perror("accept failed");
                    return 1;
                }
                puts("Connection accepted");
                
                //read from socket
                for(;;){
                    read_size = recv(client_sock , client_message , 32 , 0);
                    branchName=strtok(client_message,",");machineName=strtok(NULL,"\n");
                    puts(branchName);puts(machineName);
                    puts("hung");

                    if((inventFile=fopen("inventory.txt","r+"))==NULL){
                        perror("No database!\n");
                    }

                    while(!feof(inventFile)){
                        fgets(info,2000,inventFile);puts(info);
                        machine=strtok(info,",");
                        branch=strtok(NULL,",");
                        amount=strtok(NULL,"\n ");number=atoi(amount);number--;
                        
                        if(strcmp(machine,machineName)==0 && strcmp(branch,branchName)==0){
                            write(client_sock,success,strlen(success)+1);
                            confirm=1;
                            break;
                        }
                    }
                    
                    if(confirm==0){
                        write(client_sock,machineError,strlen(machineError)+1);
                    }
                    if(confirm==1){
                        //update database
                        sprintf(invent,"%d",number);
                        fseek(inventFile,-4,SEEK_CUR);
                        while((character=fgetc(inventFile))!='\n'){
                                if(isdigit((int)character)){
                                    fseek(inventFile,-1,SEEK_CUR);
                                    if(number==9) strcat(invent," ");
                                    fputs(invent,inventFile);
                                    break;
                                }
                        }
                        confirm=0;
                        
                    }
                    
                    if((historyFile=fopen("history.txt","r+"))==NULL){
                        perror("can't write in file");
                    }
                    //write history
                    
                        tm = *localtime(&t);
                        fseek(historyFile,0,SEEK_END);
                        fprintf(historyFile,"%s\t%s\t%d-%d-%d %d:%d:%d\n",machineName,branchName,  tm.tm_year + 1900, tm.tm_mon + 1, tm.tm_mday, tm.tm_hour, tm.tm_min, tm.tm_sec);
                        fclose(historyFile);
                        fseek(inventFile,0,SEEK_SET);
                     while(1){
                        fgets(info,2000,inventFile);puts(info);
                        if(feof(inventFile)) break;
                        machine=strtok(info,",");
                        branch=strtok(NULL,",");
                        amount=strtok(NULL,"\n ");number=atoi(amount);
                        
                        if(number<=3){
                            sprintf(invent,"%d",number+10);
                            fseek(inventFile,-4,SEEK_CUR);
                            while((character=fgetc(inventFile))!='\n'){
                                    if(isdigit((int)character)){
                                        fseek(inventFile,-1,SEEK_CUR);
                                        fprintf(inventFile,"%s",invent);
                                        fseek(inventFile,1,SEEK_CUR);
                                        printf("Delivery....\n");
                                        sprintf(client_message,"increase 10 %s in %s\n",branch,machine);
                                        send(client_sock,client_message,strlen(client_message),0);
                                        sleep(5);
                                        printf("Done!......\n");
                                        break;
                                    }
                            }
                        }
                    }

                   
                    
                    fclose(inventFile);

                    if(read_size == 0){
                        puts("Client disconnected");
                        fflush(stdout);
                    }
                    else if(read_size == -1){
                        perror("recv failed");
                    }
                }
                close(client_sock);
        }

    }

    

    
    return 0;
}

void waitFor (unsigned int secs) {
    unsigned int retTime = time(0) + secs;   // Get finishing time.
    while (time(0) < retTime);               // Loop until it arrives.
}